File: 589.py
from typing import List

# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

# 递归
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
      ans = []
      self.helper(root, ans)
      return ans

    def helper(self, node, ans):
      if node:
        ans.append(node.val)
        for ch in node.children:
          self.helper(ch, ans)

from collections import deque
# 迭代
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
      q = deque([root])
      ans = []
      while q:
        node = q.popleft()
        ans.append(node.val)
        for i in range(len(node.children)):
          q.appendleft(node.children[len(node.children) - i - 1])
      return ans

File: go/589.go
package main

func preorder(root *Node) (res []int) {
	var helper func(*Node)
	helper = func(node *Node) {
		if node == nil {
			return
		}
		res = append(res, node.Val)
		for _, child := range node.Children {
			helper(child)
		}
	}
	helper(root)
	return
}
